src/main/src/main/java/org/geoserver/ows/DisabledServiceCheck.java

```diff
@@ -63,6 +74,37 @@ public Service serviceDispatched(Request request, Service service) {
                         throw new ServiceException(
                                 "Service " + info.getName() + " is disabled", ServiceException.SERVICE_UNAVAILABLE);
                     }
+                    // check if service is disabled for layer
+                    String context = request.getContext();
+                    if (context != null && context.contains("gwc/service")) {
+                        // Account for context provided to GwcServiceDispatcherCallback
+                        if (context.contains("/gwc/service")) {
+                            context = context.substring(0, context.indexOf("/gwc/service"));
+                        } else {
+                            context = context.substring(0, context.indexOf("gwc/service"));
+                        }
+                    }
+                    if (context != null && context.contains("/")) {
+                        String layerName = context.replace("/", ":");
+                        LayerInfo layerInfo = geoServer.getCatalog().getLayerByName(layerName);
+                        if (layerInfo == null) {
+                            // DisabledServiceResourceFilter has checked getDisabledServices() list and hidden layer
+                            throw new ServiceException(
```

Comment:
Can anyone think of a way of preventing DisabledServiceResourceFilter from being applied here? If so I could look up the layerInfo and know specifically that it is disabled (rather than not found). 

This would be much better and prevent the need to do a special case for ``/gwc/service`` above.